lib/kcl.js

```diff
@@ -153,41 +154,70 @@ module.exports = function(config, kinesis) {
       return delete instanceShardList[shard.id];
     }
 
-    // expose the checkpoint function to the procerssRecords, so they can checkpoint
+    // expose the checkpoint function to the processRecords, so they can checkpoint
     // without requesting more records.
     shard.checkpointFunc = db.checkpoint.bind(null, shard);
 
-    kinesis.getRecords(
-      { ShardIterator: shard.iterator, Limit: config.limit },
-      function(err, resp) {
-        if (err && err.code === 'SyntaxError') return setTimeout(getRecords.bind(this, shard), 100);
-        if (err) throw err;
+    if (config.verbose) logger.info('[ getRecords ] params:', JSON.stringify({
+      ShardIterator: shard.iterator,
+      Limit: config.limit
+    }));
 
-        if (config.cloudwatchNamespace &&
-          resp.Records &&
-          resp.Records[0] &&
-          resp.Records[0].ApproximateArrivalTimestamp) {
-          throttledPutToCloudwatch('ShardIteratorAgeInMs', (+new Date()) - resp.Records[0].ApproximateArrivalTimestamp, 'Milliseconds', shard.id);
-        }
+    var req = kinesis.getRecords({ShardIterator: shard.iterator, Limit: config.limit});
+    var requestTimeOut = setTimeout(req.abort.bind(req), 5000);
 
-        // malformed responses will have an undefined NextShardIterator. In that case, we keep the existing one.
-        shard.iterator = resp.NextShardIterator === undefined ? shard.iterator : resp.NextShardIterator;
-        shard.lastGetRecords = +new Date();
+    req.on('error', function (err) {
+      if (err.code === 'SyntaxError' || err.code === 'RequestAbortedError') {
+        if (config.verbose) logger.info('[ getRecords ] Found SyntaxError', err);
+        return setTimeout(getRecords.bind(this, shard), 500);
+      }
+      if (err.code === 'ProvisionedThroughputExceededException') {
+        if (config.verbose) logger.info('[ getRecords ] Found ProvisionedThroughputExceededException', err);
+        return setTimeout(getRecords.bind(this, shard), 1000);
```

Comment:
@rclark we have a very low % of requests which, with the previous code, did not return an error, nor a response, and did not time out. Lowering the httpOptions and maxRetries did not help, so I built this manual retry mechanism mostly to provide better visibility into what was going on. 

The effect was that kine would hang for a full 5 minutes, and its `heartbeat` function would eventually time out and kill the process.

After changing to this method, I've seen a small number of `SyntaxError` get caught and retried – my understanding was that the request was fine, but perhaps a truncated payload could cause this to happen on parsing, which is why it wasn't retried.